### PR TITLE
Cleanup and optimization

### PR DESCRIPTION
**Pull request by spautz**
This has a rollup of several improvements discussed in passing over the past few months:
* There are now separate counts for different results of running (instead of calls and recomputations, it's now invokes (the entry) and full-run, phantom-run, skipped-run, or aborted-run (the possible exits)
* Added callbacks for the above events
* Added optimization to check the eventual root selectors before checking the immediate selectors, when determining whether or not to re-run. Since root selectors will almost always be faster, and since intermediate selectors' checks will ultimately result just check their root selectors, this skips the middleman
* Made shallowEqual the default output check: this gives better results out of the box, but makes the default here different from Reselect's default behavior
* Added test coverage for comparison presets and key-generating helpers

### CODE REVIEW

**Comment by spautz** [comment]
Oops: the build is failing due to two linter errors. Will fix later.

File: README.md [hunk 1]
```diff
@@ -180,3 +180,8 @@ performanceChecksCallback | Boolean | Gets called for every failed performanceCh
 warningsEnabled | Boolean | Will notify you about library misuse and invalid/incompatible options. (Barely implemented, mostly to-do.)
 warningsCallback | Function | Gets called for every warning item; this is `console.warn` by default.
 exceptionCallback | Function | Gets called if your selector function throws an exception.
+onInvoke | Function | Callback fired whenever the selector is executed. Useful for debugging.
+onSkippedRun | Function | Callback fired when a selector returns its cached value directly. Useful for debugging.
+onPhantomRun | Function | Callback fired when a selector runs but returns something equivalent to its cached value. Useful for debugging.
```
Comment (doelgonzo) [inline README.md:185]: I like this. I would like to be able to use this to see if a parent selector isn't optimized as it should be 👍

**Review by doelgonzo — APPROVED** [review]
Like the change and the test additions

**Comment by sam1463** [comment]
Looks good to me.

**Pull request merged**